File: test/e2e/Apps/BasicJava/src/main/java/com/function/CustomExceptionProperties.java
package com.function;

import java.time.Instant;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import com.function.exceptions.BusinessValidationException;
import com.microsoft.azure.functions.ExecutionContext;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.durabletask.FailureDetails;
import com.microsoft.durabletask.TaskFailedException;
import com.microsoft.durabletask.TaskOrchestrationContext;
import com.microsoft.durabletask.azurefunctions.DurableActivityTrigger;
import com.microsoft.durabletask.azurefunctions.DurableOrchestrationTrigger;

/**
 * Validates the custom-exception-properties feature for Durable Functions Java.
 *
 * <p>The {@code BusinessActivity} throws a {@link BusinessValidationException} carrying structured
 * properties. A registered {@link BusinessExceptionPropertiesProvider} attaches those properties to
 * the failure, which the Durable Task host extension surfaces on {@code FailureDetails.Properties}.
 * The orchestration catches the propagated failure and returns its {@link FailureDetails} so the
 * e2e test can assert the custom properties survived the round-trip. Mirrors the BasicNode and
 * BasicDotNetIsolated reference implementations.
 */
public class CustomExceptionProperties {

    /**
     * Orchestrator: OrchestrationWithCustomException.
     */
    @FunctionName("OrchestrationWithCustomException")
    public FailureDetails orchestrationWithCustomException(
            @DurableOrchestrationTrigger(name = "ctx") TaskOrchestrationContext ctx) {
        try {
            ctx.callActivity("BusinessActivity", ctx.getInstanceId(), String.class).await();
            return null;
        } catch (TaskFailedException ex) {
            return ex.getErrorDetails();
        }
    }

    /**
     * Activity: BusinessActivity - throws an exception carrying custom properties.
     */
    @FunctionName("BusinessActivity")
    public String businessActivity(
            @DurableActivityTrigger(name = "instanceId") String instanceId,
            final ExecutionContext context) throws BusinessValidationException {
        Map<String, Object> dictionaryProperty = new LinkedHashMap<>();
        dictionaryProperty.put("error_code", "VALIDATION_FAILED");
        dictionaryProperty.put("retry_count", 3);
        dictionaryProperty.put("is_critical", true);

        List<Object> listProperty = Arrays.asList("error1", "error2", 500, null);

        throw new BusinessValidationException(
                "Business logic validation failed",
                "validation-error-123",
                100,
                999999999L,
                Instant.parse("2025-10-15T14:30:00Z"),
                dictionaryProperty,
                listProperty,
                null);
    }
}

File: test/e2e/Apps/BasicNode/src/functions/CustomExceptionProperties.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import * as df from 'durable-functions';
import {
    ActivityHandler,
    ExceptionPropertiesProvider,
    OrchestrationContext,
    OrchestrationHandler,
    TaskFailedError,
} from 'durable-functions';
import { BusinessValidationException } from '../Shared/ExceptionTypes';

// Register a global provider that surfaces custom properties from thrown
// exceptions into FailureDetails.Properties. Mirrors the .NET isolated
// TestExceptionPropertiesProvider used by BasicDotNetIsolated.
const provider: ExceptionPropertiesProvider = {
    getExceptionProperties(error: unknown): Record<string, unknown> | undefined {
        if (error instanceof BusinessValidationException) {
            return {
                StringProperty: error.stringProperty,
                IntProperty: error.intProperty,
                LongProperty: error.longProperty,
                DateTimeProperty: error.dateTimeProperty,
                DictionaryProperty: error.dictionaryProperty,
                ListProperty: error.listProperty,
                NullProperty: error.nullProperty,
            };
        }
        return undefined;
    },
};
df.app.setExceptionPropertiesProvider(provider);

// Activity: BusinessActivity - throws an exception carrying custom properties.
const BusinessActivity: ActivityHandler = (): void => {
    throw new BusinessValidationException(
        "Business logic validation failed",
        "validation-error-123",
        100,
        999999999,
        "2025-10-15T14:30:00.000Z",
        {
            error_code: "VALIDATION_FAILED",
            retry_count: 3,
            is_critical: true,
        },
        ["error1", "error2", 500, null],
        null
    );
};
df.app.activity("BusinessActivity", { handler: BusinessActivity });

// Sub-orchestration: calls the activity and lets the failure (with its
// structured FailureDetails) propagate out so a parent can observe it. Used by
// the nested scenario below.
const BusinessSubOrchestration: OrchestrationHandler = function* (context: OrchestrationContext) {
    yield context.df.callActivity("BusinessActivity");
};
df.app.orchestration("BusinessSubOrchestration", BusinessSubOrchestration);

// Parent orchestration: calls the activity directly, catches the propagated
// TaskFailedError, and returns its FailureDetails. This mirrors the .NET
// isolated OrchestrationWithCustomException reference so the custom-exception-
// properties feature is validated for Node exactly as it works for dotnet
// isolated.
const OrchestrationWithCustomException: OrchestrationHandler = function* (context: OrchestrationContext) {
    try {
        yield context.df.callActivity("BusinessActivity");
    } catch (e: unknown) {
        if (e instanceof TaskFailedError) {
            return e.failureDetails;
        }
        throw e;
    }

    // Should never reach here.
    return null;
};
df.app.orchestration("OrchestrationWithCustomException", OrchestrationWithCustomException);

// Nested scenario: parent orchestration that calls a sub-orchestration (which
// in turn calls the failing activity), catches the propagated TaskFailedError,
// and returns its FailureDetails — exercising FailureDetails propagation across
// a sub-orchestration boundary.
const NestedOrchestrationWithCustomException: OrchestrationHandler = function* (
    context: OrchestrationContext
) {
    try {
        yield context.df.callSubOrchestrator("BusinessSubOrchestration");
    } catch (e: unknown) {
        if (e instanceof TaskFailedError) {
            return e.failureDetails;
        }
        throw e;
    }

    // Should never reach here.
    return null;
};
df.app.orchestration("NestedOrchestrationWithCustomException", NestedOrchestrationWithCustomException);
